Add appender clear functionality

krleonid opened the pull request:
Adds a Clear() method to the DuckDB Appender API that clears buffered data without flushing it to the database.
This allows users to discard uncommitted appends and reuse the appender instance.

The clear operation resets the appender’s internal state — including its chunk, buffered collection, and column index — without committing data to the database.
Includes tests verifying that cleared data is not flushed and that the appender remains usable after clearing.

This feature is useful when data needs to be discarded before committing.

krleonid committed:
Add appender clear functionality

krleonid committed:
Merge branch 'main' of github.com:krleonid/duckdb into appender-clear

krleonid committed:
Update appender clear tests

taniabogatsch commented:
Actually - I noticed just now that this PR includes exposing the functionality to the C API. So maybe we can add #19623 for the C++ side, and use this PR for the C API changes. WDYT @krleonid @EtgarDev?

taniabogatsch commented on src/include/duckdb/main/capi/header_generation/apis/v1/v1.2/v1.2.0.json:
```diff
@@ -288,6 +288,7 @@
         "duckdb_appender_column_type",
         "duckdb_appender_error",
         "duckdb_appender_flush",
+        "duckdb_appender_clear",
```
Instead of adding the function to the stable API, we need to first expose it only in the unstable part. Have a look at `src/include/duckdb/main/capi/header_generation/README.md` for adding functions to the unstable part, or have a look at the folder containing examples: `src/include/duckdb/main/capi/header_generation/apis/v1/unstable`.

taniabogatsch commented on test/api/capi/test_capi_appender.cpp:
```diff
@@ -1252,3 +1252,94 @@ TEST_CASE("Test upserting using the C API", "[capi]") {
+	REQUIRE(duckdb_appender_error(appender) == nullptr);
+
+	// append a bunch of values that should be cleared
+	for (idx_t i = 0; i < 2000; i++) {
```
Let's actually append a bit over 2048, to ensure we're flushing a full chunk to the column data collection.

taniabogatsch reviewed (changes requested):
Looking good! 💪 Just left a few comments - and let's eventually strip this PR down to the C API side, once we have Etgars C++ side merged? (We can wait with that ofc).

taniabogatsch commented:
> sounds great! Then, I'll merge mine, and @krleonid will only remove the commits related to the C++

Exactly!

krleonid committed:
Add duckdb_appender_clear C API function

- Added duckdb_appender_clear to appender JSON with detailed description
- Added implementation in appender-c.cpp that calls BaseAppender::Clear()
- Added duckdb_appender_clear to v1.2.0 API struct
- Added comprehensive test in test_capi_appender.cpp
- Regenerated headers using make generate-files

taniabogatsch commented:
Merged as part of https://github.com/duckdb/duckdb/pull/19623.
